### DISCUSSION TITLE
Move materialization history when moving assets from one code location to another

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#17180] [Q&A] [answered]
I moved a number of assets from one code location to another. Is it possible to move the asset materialization history as well?

**Comment by sryza** [comment, marked as answer]
As long as an asset has the same asset key, it's materialization history should be unaffected by what code location it's in. I.e. you shouldn't need to take any additional actions to migrate an asset's materialization history between code locations, if the asset's key doesn't change.

**Reply by brandonpeebles** [reply]
Hi @sryza — has there been any exploration around maintaining/migrating asset history when changing an asset key slightly? Our use case is that we're adopting Dagster in the middle of a larger data/infrastructure migration, so we pretty frequently come across situations where we're like "hmm we should maybe rename this or change the naming convention". 

Dagster handles this pretty easily, but the downside is that we immediately break the asset history and have a lot of orphaned materializations we'd need to remove (or else it's cluttered). It also ends up breaking the history of assets on the Insights feature. It's not the end of the world when it happens since hopefully we'll get to a steady state soon, but until then it makes us more likely to stick with a less ideal naming convention for keys of fear of breaking asset history.

**Reply by sryza** [reply]
We have an issue to track this but alas don't have a straightforward way to do this right now: https://github.com/dagster-io/dagster/issues/9512.

Manually reporting asset materializations can also be helpful in some cases: https://docs.dagster.io/concepts/assets/external-assets#using-the-python-api.

**Comment by alexberndt** [comment]
Note that although asset history might remain intact during a migration, you _may_ need to copy files depending on the IO manager each asset is using.

### Illustrative Example

For example, we are using the [GCSPickleIOManager](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-gcp/dagster_gcp/gcs/io_manager.py#L80) where **each code location writes to a different GCS location**:


```python
# old code location
defs = Definitions(
    ...
    resources={
        "io_manager": GCSPickleIOManager(
            gcs=gcs_resource,
            gcs_bucket="dagster-assets-gcs-bucket",
            gcs_prefix="old-code-location",
        )
   }
) 

# new code location
defs = Definitions(
    ...
    resources={
        "io_manager": GCSPickleIOManager(
            gcs=gcs_resource,
            gcs_bucket="dagster-assets-gcs-bucket",
            gcs_prefix="new-code-location",
        )
   }
) 
```

This requires users to copy `gs://dagster-assets-gcs-bucket/old-code-location/<asset-key>` to `gs://dagster-assets-gcs-bucket/new-code-location/<asset-key>` for each moved `<asset-key>`.
